### ISSUE TITLE
Command line tool to wrap release scripts

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3439]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#3439|
|Type|enhancement|
|Created|2008-09-17 02:01:22Z|
|Branch|https://github.com/twisted/twisted/tree/release-script-3439|

[twisted/python/_release.py](twisted/python/_release.py) now contains APIs for doing all of these things but there are no command line tools which wrap these them. These APIs mainly just invoke entrypoints of tools which do have command line interfaces, though. The xhtml and troff files are converted by invoking lore's main point. The book is a bit more complicated, since several steps are involved, and one of the tools is latex. It is not intended that users invoke these APIs, so where possible the documentation should not refer to them.

<details><summary>Searchable metadata</summary>

```
trac-id__3439 3439
type__enhancement enhancement
reporter__thijs thijs
priority__normal normal
milestone__ 
branch__branches_release_script_3439 branches/release-script-3439
branch_author__ 
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__ 
time__1221616882000000 1221616882000000
changetime__1226273629000000 1226273629000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Let's not fall into the trap of "we don't want to suggest it to users, so let's officially not document it".  The documentation should just clearly say that it is for developers, release managers, and possibly packagers.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

See also ReleaseAutomation.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Changing summary to reflect the fact that there should only be **one** _command line_ tool for doing the release, not several.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_4"></a>|@radix commented|
|-|-|

(In [24938]) Branching to 'release-script-3439'

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_5"></a>|@radix commented|
|-|-|

Ready for review

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra set owner to @radix|
|-|-|

1.```BuildTarballsScript``` (and possibly other newly introduced classes and/or methods) needs an @since tag

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_8"></a>|@radix removed owner|
|-|-|

I suggest that @since is unnecessary for any of this code because it's in an underscore-prefixed module; it's not a public API at all so when it was introduced is not relevant.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @radix|
|-|-|

  1. twisted/python/test/test_release.py
    1. `genVersion` should have a docstring and not include trailing whitespace
    1. `StructureAssertingMixin` should have a class docstring
    1. `assertStructure` shouldn't have trailing whitespace
    1. `test_changeAllProjectVersions` should have a docstring
    1. `test_buildAllTarballs` should have a docstring
    1. `test_buildAllTarballsEnsuresCleanCheckout` should have a docstring
    1. if you want, you could factor the common stuff out of `test_changeVersions` and `test_changeVersionsWithPrerelease`
    1. `test_defaultChangeVersionsVersionChanger` would probably be better if it called the default implementation and asserted something about its behavior
    1. same for `test_defaultBuildTarballsScriptBuilder`
  1. twisted/python/_release.py
    1. `runCommand` could document what it raises or returns
    1. `generateVersionFileData` has some trailing whitespace
    1. `DistributionBuilder` class docstring could use epytext markup for `blacklist` `@type` stuff
    1. buildAllTarballs raises `UncleanWorkingDirectory` if passed a checkout path which doesn't exist.  This isn't too big a deal right now, but it could be confusing to whoever runs the script.

The only non-trivial problem with the branch is that build-tarballs builds a sumo tarball that includes bin/admin/build-tarballs and bin/admin/change-versions.  I think we wanted to avoid releasing these files?

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_10"></a>|@radix set owner to @exarkun|
|-|-|

Thanks for the review. I've done everything you suggested save for 1.8-1.9. I also made buildAllTarballs create the destination directory when it doesn't exist: another nod to usability.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @radix|
|-|-|

Looks good.  There's still one line with trailing whitespace in `_release.py`.

Oops.  I just noticed that some of the tests fail on Windows.  Here are the build results:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/release-script-3439

Guess some skips need to be defined.  Humm.  Maybe the tests should also skip if `svn` or `svnadmin` (and whatever other external binaries the tests/code) should be skipped too.

Please fix those issues (and check the fixes on buildbot) then merge.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @radix|
|-|-|

Please take a look at the OS X failures as well.  Sorry for not mentioning those before.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_14"></a>|@radix commented|
|-|-|

(In [25225]) run commands with the default sigchld handler. refs [#3439](https://github.com/twisted/twisted/issues/3439)

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_15"></a>|@radix set owner to @exarkun|
|-|-|

Ok, I've made all the tests which invoke runCommand (directly or indirectly) install the default sigchld handler for the duration of the test. Tests pass or skip on OS X, windows, and Ubuntu.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @radix|
|-|-|

The docs for `DistributionBuilder.blacklist` end in a typo, _build_ instead of _built_.

Can you add that comment we talked about to the new scripts mentioning that they're not intended to be included in the release?

Otherwise good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_17"></a>|@radix set status to `closed`|
|-|-|

(In [25335]) Merge release-script-3439

Author: radix
Reviewer: exarkun
Fixes: [#3439](https://github.com/twisted/twisted/issues/3439)

Two new scripts, bin/admin/build-tarballs and build/admin/change-versions,
now automate most of basics of the release process. These scripts are
excluded from the release itself.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
